src/devices/src/virtio/net/device.rs

```diff
@@ -139,26 +54,123 @@ fn init_vnet_hdr(buf: &mut [u8]) {
+
+    config_space: Vec<u8>,
     guest_mac: Option<MacAddr>,
 
+    device_activated: bool,
+
+    mmds_ns: Option<MmdsNetworkStack>,
+
     #[cfg(test)]
     test_mutators: tests::TestMutators,
 }
 
-impl NetEpollHandler {
+impl Net {
+    /// Create a new virtio network device with the given TAP interface.
+    pub fn new_with_tap(
+        tap: Tap,
+        guest_mac: Option<&MacAddr>,
+        mem: GuestMemory,
+        rx_rate_limiter: RateLimiter,
+        tx_rate_limiter: RateLimiter,
+        allow_mmds_requests: bool,
+    ) -> Result<Self> {
+        // Set offload flags to match the virtio features below.
+        tap.set_offload(
+            net_gen::TUN_F_CSUM | net_gen::TUN_F_UFO | net_gen::TUN_F_TSO4 | net_gen::TUN_F_TSO6,
+        )
+            .map_err(Error::TapSetOffload)?;
+
+        let vnet_hdr_size = vnet_hdr_len() as i32;
+        tap.set_vnet_hdr_size(vnet_hdr_size)
+            .map_err(Error::TapSetVnetHdrSize)?;
+
+        let mut avail_features = 1 << VIRTIO_NET_F_GUEST_CSUM
+            | 1 << VIRTIO_NET_F_CSUM
+            | 1 << VIRTIO_NET_F_GUEST_TSO4
+            | 1 << VIRTIO_NET_F_GUEST_UFO
+            | 1 << VIRTIO_NET_F_HOST_TSO4
+            | 1 << VIRTIO_NET_F_HOST_UFO
+            | 1 << VIRTIO_F_VERSION_1;
+
+        let mut config_space;
+        if let Some(mac) = guest_mac {
+            config_space = vec![0; MAC_ADDR_LEN];
+            config_space[..].copy_from_slice(mac.get_bytes());
+            // When this feature isn't available, the driver generates a random MAC address.
+            // Otherwise, it should attempt to read the device MAC address from the config space.
+            avail_features |= 1 << VIRTIO_NET_F_MAC;
+        } else {
+            config_space = Vec::new();
+        }
+
+        let mut queue_evts = Vec::new();
+        for _ in QUEUE_SIZES.iter() {
+            queue_evts.push(EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?);
+        }
+
+        let queues = QUEUE_SIZES.iter().map(|&s| Queue::new(s)).collect();
+
+        let mmds_ns = if allow_mmds_requests {
+            Some(MmdsNetworkStack::new_with_defaults())
+        } else {
+            None
+        };
+
+        Ok(Net {
+            tap,
+            avail_features,
+            acked_features: 0u64,
+            mem,
+            queues,
+            queue_evts,
+            rx_rate_limiter,
+            tx_rate_limiter,
+            rx_deferred_frame: false,
+            rx_deferred_irqs: false,
+            rx_bytes_read: 0,
+            rx_frame_buf: [0u8; MAX_BUFFER_SIZE],
+            tx_frame_buf: [0u8; MAX_BUFFER_SIZE],
+            tx_iovec: Vec::with_capacity(QUEUE_SIZE as usize),
+            interrupt_status: Arc::new(AtomicUsize::new(0)),
+            interrupt_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            guest_mac: Some(MacAddr::from_bytes_unchecked(
+                &config_space.clone()[..MAC_ADDR_LEN],
+            )),
```

Comment:
I believe this will crash for net devices with no configured MAC address.

In such a case, `config_space` is an empty vec, and here you're cloning it and addressing out of bounds.

Reply:
Yes, you are right. I have to check first if the config_space is populated.